### DISCUSSION TITLE
"The control already has a visual parent" with ItemsRepeater

### DISCUSSION THREAD

**Discussion by JGKle** [AvaloniaUI/Avalonia#14005] [Q&A]
I'm getting the dreaded "The control already has a visual parent" error when using an ItemsRepeater whose ItemsSource is bound to a property that gets populated from xaml.  Details:

My app uses a NavigationView-based UI (using FluentAvalonia). Each page has a PageHeader control, which the page title, some global action buttons, & some page-specific action buttons:

```

<UserControl .....  x:Class="xxx.PageHeader" x:Name="Header">
  
  <Grid Height="48" Margin="50 0 10 0" >

    <TextBlock Text="{Binding PageName, ElementName=Header}" />

    <!--Action Buttons-->
    <StackPanel Orientation="Horizontal" HorizontalAlignment="Right">

      <!--Page-specific action buttons-->
      <ItemsRepeater ItemsSource="{Binding PageActions, ElementName=Header}">
        <ItemsRepeater.ItemTemplate>
          <DataTemplate x:DataType="{x:Type Control}">
            <ContentControl Content="{Binding}" Margin="0 0 10 0" />
          </DataTemplate>
        </ItemsRepeater.ItemTemplate>
      </ItemsRepeater>

      <!--Global Action Buttons-->
	  ......
    </StackPanel>

  </Grid>
</UserControl>
```

PageActions is an ObservableCollection of Controls:

```
public partial class PageHeader : UserControl
{
    public PageHeader()
    {
        InitializeComponent();
    }

    public static readonly StyledProperty<ObservableCollection<Control>> PageActionsProperty =
        AvaloniaProperty.Register<PageHeader, ObservableCollection<Control>>(nameof(PageActions), defaultValue: new ObservableCollection<Control>());
    public ObservableCollection<Control> PageActions
    {
        get => GetValue(PageActionsProperty);
        set => SetValue(PageActionsProperty, value);
    }

    //...other properties...
}
```
We can then use this in each page, specifying any buttons specific to that page:
```
<local:PageHeader PageName="{Binding PageName}">
  <local:PageHeader.PageActions>
    
     <Button Content="Page Specific Button"
                   Command="{Binding Path=((local:LogPageViewModel)DataContext).SendLogsCommand, ElementName=LogsPage}" />
  </local:PageHeader.PageActions>
</local:PageHeader>
```

This all works. I launch the app, navigate to the "Logs" page (in this case), and see the button appear in the header. I can use the button & it functions. However, as soon as I navigate *away* from the Logs page:

```
This exception was originally thrown at this call stack:
    Avalonia.Visual.ValidateVisualChild(Avalonia.Visual)
    Avalonia.Collections.AvaloniaList<T>.Add(T)
    Avalonia.Controls.Presenters.ContentPresenter.UpdateChild(object)
    Avalonia.Layout.Layoutable.MeasureCore(Avalonia.Size)
    Avalonia.Layout.Layoutable.Measure(Avalonia.Size)
    Avalonia.Layout.Layoutable.MeasureOverride(Avalonia.Size)
    Avalonia.Layout.Layoutable.MeasureCore(Avalonia.Size)
    Avalonia.Layout.Layoutable.Measure(Avalonia.Size)
    Avalonia.Layout.StackLayout.GetAverageElementSize(Avalonia.Size, Avalonia.Layout.VirtualizingLayoutContext, Avalonia.Layout.StackLayoutState)
    Avalonia.Layout.StackLayout.GetAnchorForRealizationRect(Avalonia.Size, Avalonia.Layout.VirtualizingLayoutContext)
    ...
    [Call Stack Truncated]
```

This only occurs when there are buttons added to PageActions. They work & appear as intended, but I can never leave that navigation page again - or it will crash.

I've searched & found many bug reports related to this error message, many of which are old/closed, or suggest setting `"VirtualizationMode="None"` (which no longer exists), or have no answers (i.e. https://github.com/AvaloniaUI/Avalonia/issues/13133, https://github.com/AvaloniaUI/Avalonia/discussions/11542). I'm not sure if this is the same issue or different...or more importantly, how I can solve it.

**Comment by jp2masa** [comment]
This error is caused by reusing control instances. Although detaching from a tree and attaching to a new tree should be supported, it often leads to problems.

Most Avalonia devs probably use MVVM, where the view model instances can be reused, and the data templates generate a new control when needed. It's almost impossible to get this error with MVVM.

If you don't want to use MVVM, a possible solution is to use a collection of templates instead of controls.

**Comment by JGKle** [comment]
Thanks for the reply.

I actually am using MVVM everywhere else in my app - this is the only control where I don't have a ViewModel, as I'm not super experienced & just couldn't figure out a way to do the above, where the buttons can be described in xaml. i.e. so I could say:

```
<local:PageHeader PageName="{Binding PageName}">
  <local:PageHeader.PageActions>
     <Button Content="Page Specific Button 1" />
     <Button Content="Page Specific Button 2 />
  </local:PageHeader.PageActions>
</local:PageHeader>
```

Like to do this with MVVM, I thought I'd have to do `<local:PageHeader PageActions={Binding PageActionButtons} />`, which would mean creating the buttons in the ViewModel & setting them to PageActionButtons, no?

Apologies if it's a dumb/beginner question - but if there's a better way to do it, I'd be super appreciative of a short example :)

**Comment by jp2masa** [comment]
Example:

```cs
class PageActionViewModel : ViewModelBase
{
    public string Name { get; }
    public ICommand ActionCommand { get; }
    ...
}
```

```cs
class PageViewModel : ViewModelBase
{
    public ReadOnlyObservableCollection<PageActionViewModel> PageActions { get; }
    ...
}
```

You have 2 options for the template:

1. If you use a view resolver, you just need to create the corresponding view (and register it in the resolver):
    ```xaml
    <UserControl x:Class="MyApp.Views.PageActionView" ...>
        <Button Content="{Binding Name}" Command="{Binding ActionCommand}" />
    </UserControl>
    ```
    Then bind the `ItemsSource` as you did, and the template should be resolved automatically.

2. You can create the template:
    ```xaml
        <ItemsControl ItemsSource="{Binding PageActions}">
            <ItemsControl.ItemTemplate>
                <DataTemplate>
                    <Button Content="{Binding Name}" Command="{Binding ActionCommand}" />
                </DataTemplate>
            </ItemsControl.ItemTemplate>
        </ItemsControl>
    ```

If you want to have multiple templates for page actions, the view resolver approach doesn't need any change, but the other approach would have to be replaced by `DataTemplates` or some other mechanism, depending on your use case.

**Reply by JGKle** [reply]
Thanks for the thoughtful reply. 

I think maybe I was a bit unclear tho regarding: "_a way to do the above, where the buttons can be described in xaml._" What I meant is, basically each Page has a PageHeader, but I'd like to be able to just put whatever controls I need to in the PageHeader - from the Page's xaml. i.e.:

```
<local:PageHeader PageName="{Binding PageName}">
  <local:PageHeader.PageActions>
     <Button Content="Action 1" Command="..." />

     <ToggleButton Content="Action 2" IsChecked="..." IsEnabled="..." i:Attached.Icon="..." />

     <i:Icon ... />

     <!--....any other controls--->

  </local:PageHeader.PageActions>
</local:PageHeader>

<!--...Rest of the page-->
```

Unless I'm misunderstanding, with the MVVM approach above, I'd basically need a separate ViewModel for each type of control I want to support in the PageHeader (i.e. `PageActionViewModelButton`, `PageActionViewModelToggleButton`, etc), so the DataTemplate can select the intended control based on the `x:Type` of the ViewModel. Furthermore, each ViewModel would need a property for every possible property in the underlying control I want to access (i.e. in your example, you made `Content` and `Command` - I'd need ViewModel properties for `IsEnabled`, `IsChecked`, `Icon`, etc ).

Does that make sense? Is there not a better way?

**Reply by jp2masa** [reply]
Yes, the idea is to have a viewmodel for each type of control. There are certainly other ways to achieve what you want, but I usually structure the code like this, mainly because of the following advantages:

- No need to touch view code (i.e. coupled to UI framework) to change the page actions for a specific page
    - This also makes it easier to write tests
- If you want to change the view for some kind of page action, you just need to update the XAML in one place
    - It's also easy to change templates according to the host navigation view (e.g. a simple navigation view may use `Button`s, while a more complex one may use a special button control with specific properties and behaviors that integrate it more seamlessly with the host)

**Reply by JGKle** [reply]
Alright, fair enough - maybe I'm overthinking it. It just felt a bit redundant to basically be "redefining" or "mirroring" everything in the UI framework (i.e. mirror each control I want to use with a viewmodel, mirror each property I want to use with a ViewModel property), but it definitely works, & if that's not bad practice / the preferred method, I'll defer to your expertise :)

Than much again for your help!

**Reply by JGKle** [reply]
Ok really sorry, I'm still not seeing how I can get this to work for different types of controls :/  I thought I could do like the following:

```
<ItemsControl ItemsSource="{Binding $parent.((local:PageViewModelBase)DataContext).PageActions}">
  <ItemsControl.ItemTemplate>
    
    <DataTemplate x:DataType="{x:Type local:PageHeaderActionViewModel}">
      <Button Command="{Binding ActionCommand}"
              IsVisible="{Binding IsVisible}">
        <ic:SymbolIcon Symbol="{Binding Icon}" />
      </Button>
    </DataTemplate>

    <DataTemplate x:DataType="{x:Type local:PageHeaderActionToggleViewModel}">
      <ToggleButton Command="{Binding ActionCommand}"
              IsVisible="{Binding IsVisible}"
              IsChecked="{Binding IsChecked}" >
        <ic:SymbolIcon Symbol="{Binding Icon}" />
      </ToggleButton>
    </DataTemplate>
  </ItemsControl.ItemTemplate>
</ItemsControl>
```

...So if the item is a `PageHeaderActionViewModel`, it will be a `Button`, and if it's a `PageHeaderActionToggleViewModel` it will be a `ToggleButton`. But `Unable to find a setter that allows multiple assignments to the property ItemTemplate of type Avalonia.Controls:Avalonia.Controls.ItemsControl`. I guess I still misunderstood how to accomplish this...?

**Reply by jp2masa** [reply]
Try `<ItemsControl.DataTemplates>` instead of `<ItemsControl.ItemTemplate>`. The main difference is that those templates can now be resolved for descendants, but it doesn't matter in most cases.

**Reply by JGKle** [reply]
That worked, thanks!

**Comment by timunie** [comment]
@metal450 controls should not be used as the ItemsSource. Also UserControl is not the best choice here. Consider to inherit ContentControl. Check Avalonia.Samples RatingControl for a tutorial

**Reply by JGKle** [reply]
Thanks - that's a super well-explained example! However, the missing piece I still can't seem to figure out is how to have it accept other controls, per-page - I think I explained it a bit better with an example in my follow-up comment above? https://github.com/AvaloniaUI/Avalonia/discussions/14005#discussioncomment-7923482?

Thanks so much again for the helpful replies :)
